DOCS-417: Add 28BYJ-48 stepper motor model (#958)
viamrobotics/docs: docs/components/motor/28byj-48.md

---
title: "Configure a 28BYJ-48 stepper motor"
linkTitle: "28BYJ-48"
weight: 22
type: "docs"
description: "Configure a stepper motor driven by a ULN2003 driver."
# SMEs: Rand, James
---

> The `28byj48` model of the motor component supports small [stepper motors](https://en.wikipedia.org/wiki/Stepper_motor) controlled by stepper motor drivers like [ULN2003](https://www.ti.com/product/ULN2003A) that take coil input through GPIO and move the motor when they receive a step sequence.

Reviewer: This is better, I think. But I still think it'd be helpful to add a sentence that I don't have the knowledge to write but maybe @susmitaSanyal does. Something like "If you have a stepper motor driver that <insert accurate words about what is different about gpiostepper>, see the [`gpiostepper`](../gpiostepper/) documentation." Does it have something to do with 4 wires between the board and the driver, instead of 2 for gpiostepper?

Reviewer 2: 28byj48 motor functions a little differently than the gpiostepper motor that we have a driver for. 
you can find the details to how this particular motor works[ here ](https://lastminuteengineers.com/28byj48-stepper-motor-arduino-tutorial/) 
This motor is pretty different from gpiostepper in RDK.

Author: ```suggestion
The `28byj48` model of the motor component supports small unipolar [stepper motors](https://en.wikipedia.org/wiki/Stepper_motor) controlled by stepper motor drivers like [ULN2003](https://www.ti.com/product/ULN2003A). The `28byj48` is often used for low-current and low-precision applications and supports full, half, and quarter stepping with 2048 steps in a rotation in full-step mode. If you require a higher step resolution, use the [`gpiostepper`](../gpiostepper) model.
```

> The `28byj48` model of the motor component supports [stepper motors](https://en.wikipedia.org/wiki/Stepper_motor) controlled by basic stepper driver chips (such as [ULN2003](https://www.ti.com/product/DRV8825)) that take coil input through GPIO and move the motor when they receive a step sequence.

Author: ```suggestion
The `28byj48` model of the motor component supports small [stepper motors](https://en.wikipedia.org/wiki/Stepper_motor) controlled by stepper motor drivers like [ULN2003](https://www.ti.com/product/ULN2003A) that take coil input through GPIO and move the motor when they receive a step sequence.
```

        "board": <board_name>,
        "pins": {
          "in1": <board pin number (string) of in1 pin>,
          "in2": <board pin number (string) of in2 pin>,
          "in3": <board pin number (string) of in3 pin>,
          "in4": <board pin number (string) of in4 pin>
        },
        "ticks_per_rotation": <int>
      },
      "depends_on": []
    }
  ]
}
```

{{% /tab %}}
{{% tab name="JSON Example" %}}

Example configuration for a `28byj48` stepper motor:

```json
{
  "components": [
    {
      "name": "example-board",
      "type": "board",
      "model": "pi"
    },
    {
      "name": "example-motor",
      "type": "motor",
      "model": "28byj48",
      "attributes": {
        "board": "example-board",
        "pins": {
          "in1": "16",
          "in2": "18",
          "in3": "29",
          "in4": "22"
        },
        "ticks_per_rotation": 200
      }
    }
  ]
}
```

{{% /tab %}}
{{< /tabs >}}

### Required Attributes

Name | Type | Description
-------------- | ---- | ---------------
`board` | string | Should match name of board to which the motor driver is wired.
`pins` | object | A json object containing `in1`, `in2`, `in3`, and `in4` pin numbers.

> `ticks_per_rotation` | integer | Number of full steps in a rotation. 200 (equivalent to 1.8 degrees per step) is very common. If your data sheet specifies this in terms of degrees per step, divide 360 by that number to get ticks per rotation.

Reviewer: Oooh that is possibly a good thing to mention RE what kind of motor and driver this is for? Since that's different from "standard" stepper motors?

> `ticks_per_rotation` | integer | Number of full steps in a rotation. 200 (equivalent to 1.8 degrees per step) is very common. If your data sheet specifies this in terms of degrees per step, divide 360 by that number to get ticks per rotation.

Author: ```suggestion
`ticks_per_rotation` | integer | Number of full steps in a rotation. The motor takes 5.625*(1/64)° per step, 4096 steps is 360° . So one full rotation is 4096 steps.

```

> `ticks_per_rotation` | integer | Number of full steps in a rotation. 200 (equivalent to 1.8 degrees per step) is very common. If your data sheet specifies this in terms of degrees per step, divide 360 by that number to get ticks per rotation.

Reviewer 2: This is specifically for 28byj48 motor. The motor has a gear ratio of 64:1. which is  how we were able to calculate 4096 steps per revolution. This may be different for other stepper motors depending on what kind of gear ratio they have.

        "board": <board_name>,
        "pins": {
          "in1": <board pin number (string) of in1 pin>,
          "in2": <board pin number (string) of in2 pin>,
          "in3": <board pin number (string) of in3 pin>,
          "in4": <board pin number (string) of in4 pin>
        },
        "ticks_per_rotation": <int>
      },
      "depends_on": []
    }
  ]
}
```

{{% /tab %}}
{{% tab name="JSON Example" %}}

Example configuration for a `28byj48` stepper motor:

```json
{
  "components": [
    {
      "name": "example-board",
      "type": "board",
      "model": "pi"
    },
    {
      "name": "example-motor",
      "type": "motor",
      "model": "28byj48",
      "attributes": {
        "board": "example-board",
        "pins": {
          "in1": "16",
          "in2": "18",
          "in3": "29",
          "in4": "22"
        },
        "ticks_per_rotation": 200
      }
    }
  ]
}
```

{{% /tab %}}
{{< /tabs >}}

### Required Attributes

Name | Type | Description
-------------- | ---- | ---------------
`board` | string | Should match name of board to which the motor driver is wired.
`pins` | object | A json object containing `in1`, `in2`, `in3`, and `in4` pin numbers.

> `ticks_per_rotation` | integer | Number of full steps in a rotation. The motor takes 5.625*(1/64)° per step, 4096 steps is 360° . So one full rotation is 4096 steps.

Author: ```suggestion
`ticks_per_rotation` | integer | Number of full steps in a rotation. The motor takes 5.625*(1/64)° per step, 4096 steps is 360° . One full rotation is 4096 steps.

```
